src/Compilers/Core/Portable/Serialization/StreamObjectReader.cs

```diff
@@ -0,0 +1,1020 @@
+            }
+
+            throw ExceptionUtilities.UnexpectedValue(enumType);
+        }
+
+        private Variant ReadObject(EncodingKind kind)
+        {
+            switch (kind)
+            {
+                case EncodingKind.ObjectRef:
+                    return Variant.FromObject(_referenceMap.GetValue(_reader.ReadInt32()));
+                case EncodingKind.ObjectRef_B:
+                    return Variant.FromObject(_referenceMap.GetValue(_reader.ReadByte()));
+                case EncodingKind.ObjectRef_S:
+                    return Variant.FromObject(_referenceMap.GetValue(_reader.ReadUInt16()));
+
+                case EncodingKind.Object_W:
+                    int id = _referenceMap.GetNextReferenceId();
+
+                    Type type = this.ReadType();
+                    uint memberCount = this.ReadCompressedUInt();
+
+                    if (_binder == null)
+                    {
+                        throw NoBinderException(type.FullName);
+                    }
+
+                    var reader = _binder.GetReader(type);
+                    if (reader == null)
+                    {
+                        throw NoReaderException(type.FullName);
+                    }
+
+                    if (memberCount == 0)
+                    {
+                        return ConstructObject(type, (int)memberCount, reader, id);
+                    }
+                    else
+                    {
+                        _consumerStack.Push(new Consumer(type, (int)memberCount, _valueStack.Count, reader, id));
+                        return Variant.None;
+                    }
+
+                default:
+                    throw ExceptionUtilities.UnexpectedValue(kind);
+            }
+        }
+
+        private Variant ConstructObject(Type type, int memberCount, Func<ObjectReader, object> reader, int id)
+        {
+            _memberReader.Reset();
+
+            // take members from the stack
+            for (int i = 0; i < memberCount; i++)
+            {
+                _memberReader.List.Add(_valueStack.Pop());
+            }
+
+            // reverse list so that first member to be read is first
+            _memberReader.List.Reverse();
+
+            // invoke the deserialization constructor to create instance and read & assign members           
+            var instance = reader(_memberReader);
+
+            if (_memberReader.Position != memberCount)
+            {
+                throw new InvalidOperationException($"Deserialization constructor for '{type.Name}' was expected to read {memberCount} values, but read {_memberReader.Position} values instead.");
+            }
+
+            _referenceMap.AddValue(id, instance);
+
+            return Variant.FromObject(instance);
+        }
+
+        private static Exception NoBinderException(string typeName)
+        {
+#if COMPILERCORE
+            return new InvalidOperationException(string.Format(CodeAnalysisResources.NoBinderException, typeName));
```

Comment:
Easier would be to just update CodeAnalysisResources to use the same identifier that the Workspace uses.  On the plus side, tehn you can tell exactly what the error message is, and you can see precisely what placeholders it has.